### PR TITLE
Index improvements

### PR DESCRIPTION
**Pull request by Marri**
Get indexes to the point where they are a semi-usable feature, rather than an in progress thing no one knows or cares about.

- [ ] Do we want to disallow repeating posts in an index or index section? There's no validations, so leaving it for now. If we do disallow, we need validations and also api/posts should filter duplicates in index_posts#new.
- [ ] Do we want to allow changing the IndexSection for arbitrary IndexPosts? Right now it just shows/links.

### CODE REVIEW

**Comment by Throne3d** [comment]
> Do we want to disallow repeating posts in an index or index section? There's no validations, so leaving it for now. If we do disallow, we need validations and also api/posts should filter duplicates in index_posts#new.

Within a section, I'd think it makes sense to forbid. Within an overall index, not sure! Since it's a bit awkward and doesn't really add much for us from a development perspective, I think we can just leave it as is, though.

>  Do we want to allow changing the IndexSection for arbitrary IndexPosts? Right now it just shows/links.

What does this mean? Moving a post around between sections in an index seems fine – is there something more to it than that?

File: app/models/index_section.rb [hunk 1]
```diff
@@ -2,15 +2,23 @@ class IndexSection < ApplicationRecord
   include Orderable
 
   belongs_to :index, inverse_of: :index_sections, optional: false
-  has_many :index_posts, inverse_of: :index_section, dependent: :destroy
-  has_many :posts, through: :index_posts, dependent: :destroy
+  has_many :index_posts, inverse_of: :index_section, dependent: false # This is handled in callbacks
```
Comment (Throne3d) [inline app/models/index_section.rb]: Is there a reason not to use `dependent: :nullify` here? (Wanting to audit when this changes?)

File: app/views/boards/_reorder.haml [hunk 2]
```diff
@@ -0,0 +1,51 @@
+- reset_cycle
+#reorder-sections-table
+  .content-header
+    = content_for :reorder_title
+    #loading.float-right.hidden= loading_tag # TODO duplicate id
+    #saveerror.float-right.hidden
+      - # TODO duplicate id
+      = image_tag "icons/exclamation.png", title: 'Error', class: 'vmid', alt: '!'
```
Comment (Throne3d) [inline app/views/boards/_reorder.haml]: `alt`s should be changed in line with #1242 once that gets merged (in this case, it should be empty)

File: app/views/boards/_reorder.haml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  %ul.sortable.table-list
+    - sections.each do |section|
+      %li.padding-5.section-ordered{class: cycle('even', 'odd'), data: {id: section.id, type: section.class.to_s, order: section.section_order}}
+        .section-ordered-handle
+          = image_tag "icons/arrow_double.png", title: 'Reorder', class: 'disabled-arrow', alt: '↕'
+        = link_to section.name, section
+        .float-right
+          = image_tag "icons/arrow_up.png", class: "section-up disabled-arrow"
+          = image_tag "icons/arrow_down.png", class: "section-down disabled-arrow"
+          = link_to url_for(section) + "/edit" do # TODO
```
Comment (Throne3d) [inline app/views/boards/_reorder.haml]: What's TODO here? `edit_section_url(section)` or something?

File: app/views/boards/_reorder.haml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        Saved
+    %ul.sortable.table-list
+      - unsectioned_posts.each do |post|
+        %li.section-ordered{class: cycle('even', 'odd'), data: {id: post.id, type: post.class.to_s, order: post.section_order}}
+          .section-ordered-handle
+            = image_tag "icons/arrow_double.png", title: 'Reorder', class: 'disabled-arrow', alt: '↕'
+          = link_to post.subject, post_path(post)
+          .float-right
+            = image_tag "icons/arrow_up.png", class: "section-up disabled-arrow"
+            = image_tag "icons/arrow_down.png", class: "section-down disabled-arrow"
```
Comment (Throne3d) [inline app/views/boards/_reorder.haml]: (missing trailing newline)

File: app/views/index_sections/_single.haml [hunk 5]
```diff
@@ -1,25 +1,25 @@
...
-- section_posts = posts_from_relation(section.posts.ordered_by_index, with_pagination: !in_collection, select: ', index_posts.description as index_description')
+    %th.subber.index-description{colspan: 7}= sanitize_written_content(section.description)
+- section_posts = posts_from_relation(section.posts.ordered_by_index, with_pagination: !in_collection, select: ', index_posts.description as index_description, index_posts.id as index_post_id')
 - if section_posts.present?
-  = render partial: 'posts/list_item', collection: section_posts, as: :post
+  = render partial: 'posts/list_item', collection: section_posts, as: :post, locals: {index: section.index}
 - else
   %tr
-    %td.centered.padding-10.no-posts{ class: cycle('even', 'odd'), colspan: 6 } — No posts yet —
+    %td.centered.padding-10.no-posts{ class: cycle('even', 'odd'), colspan: 7 } — No posts yet —
```
Comment (Throne3d) [inline app/views/index_sections/_single.haml]: small: We should probably extract colspan to a variable so future diffs are more readable

File: app/views/boards/_reorder.haml (hunk 3 above)
Comment (Marri) [reply app/views/boards/_reorder.haml]: Problem is that it's either edit_board_section_url or edit_index_post_url; url_for will give the correct thing for the model but there's no equivalent for the edit url and appending "/edit" is not exactly Good Form

Comment (teceler) [reply app/views/boards/_reorder.haml]: use a helper and check the class?

Comment (Throne3d) [reply app/views/boards/_reorder.haml]: `url_for(section, action: 'edit')`?

Comment (Throne3d) [reply app/views/boards/_reorder.haml]: or even probably `link_to section, action: 'edit'`, since I think it defers to url_for directly

Comment (Throne3d) [reply app/views/boards/_reorder.haml]: If that doesn't work, it looks like `edit_polymorphic_url(section)` may? https://api.rubyonrails.org/classes/ActionDispatch/Routing/PolymorphicRoutes.html

File: app/models/index_section.rb (hunk 1 above)
Comment (Marri) [reply app/models/index_section.rb]: There is indeed a reason, and it is that we need the callbacks to handle reordering them

File: app/views/boards/_reorder.haml (hunk 3 above)
Comment (Marri) [reply app/views/boards/_reorder.haml]: the action ones didn't work but edit_polymorphic did

File: app/views/boards/_reorder.haml (hunk 2 above)
Comment (Marri) [reply app/views/boards/_reorder.haml]: Did get to go through merge conflict fun, yep xD

**Pull request merged**